### PR TITLE
WIP:Django upgrade to 2.2

### PR DESCRIPTION
**Pull request by bramalingam**
--exclude

This is still work in progress, opening a PR to collect feedback and push changes accordingly.
Doing it serially to start with from the current version to 2.0 and then potentially to 2.2

Trying to not touch the regex's as of now, wanted to the transition to using converter specifications in the url pattern but have avoided it for the time being. 

Happy to take feedback and contribute accordingly.
@will-moore @joshmoore

### CODE REVIEW

**Comment by will-moore** [comment]
Hi Balaji,
Great that you're able to help with this!
Have you looked at the work I started on this a while back? https://github.com/will-moore/omero-web/tree/django_22

I haven't worked on it for a little while, so the branch is out of date and needs origin/master merging back in or rebasing.
But it looks like you're going through the same steps as me, so it might save you some work.

I certainly prefer the new `path()` syntax compared with using regexes, although I found we still need regexes for some cases.

 Cheers!
  Will.

**Comment by bramalingam** [comment]
Hi Will,

I saw the PR, thank you 
I am just following the official Django documentation and release notes and running tests before every commit,
https://docs.djangoproject.com/en/3.1/howto/upgrade-version/

thus following the same path as you (to do sanity checks after every step). I will eventually be moving all your commits as well.
This process just helped me to get to the bottom of the release notes (and understand the choices you have made as well).

I started with using `path` as well,
https://github.com/bramalingam/omero-web/tree/django_upgrade
but it meant touching most of the regex and it looked like a lot of breaking changes and I wasn't sure about our testing coverage as it stands (for all url-paths).

If the target is to move towards a stable version of 2.2 quickly (re_path usage looked like the least breaking path). 

Happy to revert back to `path` because it improves readability of the URL patterns and provides type safety as well with <int:objectid>. Will try and push commits for one module at a time in that case (to make reviewing easier), if you prefer this route. Please let me know either way.

**Comment by bramalingam** [comment]
@will-moore : I have replaced the usage of re_path with path in the most obvious cases.
All urls with optional arguments still use re_path. 
This currently improves readability a bit. 

Will push the following this week,
1) Cherry-pick the rest of your commits from your PR.
2) Upgrade to using version 2.2
3) Fix Flake8 issues.
4) Integrate with 'django_extensions'

I am also going to look into writing some unit tests to test the minimum,
`assertEqual(200, response.status_code)`
for all the urls that I have touched (if they are missing). 
Happy to take some feedback on this front, as well.

**Comment by will-moore** [comment]
Hi Balaji,
 Almost all the view methods are wrapped with the `@login_required` decorator which will return a 403 (not logged in) unless you are connected to OMERO. So for unit tests I think you'll be restricted to testing for that. That won't tell you much, but it will at least tell you it's not a 404.

**Comment by bramalingam** [comment]
@will-moore : I have cherry-picked your commits into this branch and started testing the URL's manually to start with.

I identified a primary issue with 2.2 upgrade being the following,
https://docs.djangoproject.com/en/3.1/releases/2.1/
specifically: url()’s support for inline flags in regular expression groups ((?i), (?L), (?m), (?s), and (?u)) is removed

and I found this following PR,
https://github.com/ome/omero-web/pull/93

Looks like you had started some work on this front as well. Will try and resolve this today.

@joshmoore : Meanwhile, it looks like black formatting PR has gone in,
Any suggestions on a nice way to rebase this PR?

**Comment by bramalingam** [comment]
@will-moore : Reg case insensitive urls, here are the 3 options that I could find,

1. Dropping support for case insensitive urls (am guessing this is an extreme and you wouldn't want this).
2. Using a middleware (am guessing you might not be in favor of this option as well, given its performance implications).
3. Registering a custom path converter
[](https://docs.djangoproject.com/en/3.1/topics/http/urls/#registering-custom-path-converters. )
This looks like the most viable option, happy to take your feedback on the same.

Thought I'd take some guidance before proceeding with the implementation.

**Comment by bramalingam** [comment]
@will-moore : Reg case insensitive urls, here are the 3 options that I could find,

1) Dropping support for case insensitive urls (am guessing this is an extreme and you wouldn't want this).
2) Registering a custom path converter
https://docs.djangoproject.com/en/2.2/topics/http/urls/#registering-custom-path-converters
This looks like the most viable option according to me, happy to take your feedback on the same.
3) Using a middleware (given its performance implications, this might not be a great option as well).

Thought I'd take some guidance before proceeding with the implementation.

**Comment by joshmoore** [comment]
> Any suggestions on a nice way to rebase this PR?

If you can either squash all the commits then run pre-commit on the sum or you can try to `rebase -i` running pre-commit on each commit.

**Comment by manics** [comment]
I _think_ running pre-commit on your branch, then merging in `master` and resolving conflicts, should also work.

**Comment by bramalingam** [comment]
Thank you @joshmoore and @manics : will try and update

**Comment by will-moore** [comment]
I tried something last night - see https://github.com/will-moore/omero-web/tree/django_upgrade_re_path
That tries to merge everything, with black. But I haven't tried running it yet!

**Comment by bramalingam** [comment]
Hi Will, given the issue with (?i) being there in our urls to generate case insensitive urls the current branch is still not ready to run. I will try and push a change with the converters and test your branch as well, on first look it looks perfect (thanks for fixing the default value for handler404 as well).

File: omeroweb/urls.py [hunk 1]
```diff
@@ -120,12 +120,12 @@ def redirect_urlpatterns():
...
-    re_path(r"^(?i)webclient/", include("omeroweb.webclient.urls")),
-    re_path(r"^(?i)url/", include("omeroweb.webredirect.urls")),
-    re_path(r"^(?i)feedback/", include("omeroweb.feedback.urls")),
-    re_path(r"^(?i)api/", include("omeroweb.api.urls")),
+    path(r"webgateway/", include("omeroweb.webgateway.urls")),
+    path(r"webadmin/", include("omeroweb.webadmin.urls")),
+    path(r"webclient/", include("omeroweb.webclient.urls")),
+    path(r"url/", include("omeroweb.webredirect.urls")),
+    path(r"feedback/", include("omeroweb.feedback.urls")),
+    path(r"api/", include("omeroweb.api.urls")),
```
Comment (bramalingam) [inline omeroweb/urls.py:128]: @joshmoore @will-moore : I've gone for a breaking solution after having reviewed a lot of Django docs, dropping support for case insensitive urls sounded like the correct thing to do. Happy to add variations of urls to the list if required. Would be great to get some feedback on this front.

Comment (bramalingam) [reply omeroweb/urls.py:128]: If you are happy with this decision, I'll try and add some tests and wrap up this PR. I have tested a few urls manually, but I'd like to look into the unit-tests and concentrate on adding as many unit tests as possible

Comment (manics) [reply omeroweb/urls.py:128]: It sounds sensible to me, looking at a randomly chosen web app https://github.com/ome/omero-iviewer/blob/v0.10.1/plugin/omero_iviewer/urls.py it looks like the URLs are already case sensitive

Comment (bramalingam) [reply omeroweb/urls.py:128]: Thank you for the feedback @manics . Will look into adding some unit tests

**Comment by will-moore** [comment]
Just starting to test locally, I realise that all the `omero.web.apps` also need to work with Django 2.2.
Principle ones are `omero-figure` and `omero-iviewer`. I'm going to remove all these apps initially for testing, but then we need to look at upgrading the various apps.
For my own records, I currently have these locally:
```
$ omero config get omero.web.apps ["omero_iviewer", "parade_crossfilter", "omero_figure", "corsheaders", "omero_parade", "omero_webtest", "my_channel_viewer", "omero_mapr", "omero_openwith_url"]
```

**Comment by will-moore** [comment]
@bramalingam Travis is failing:

```
black....................................................................Failed
- hook id: black
- files were modified by this hook
reformatted /home/travis/build/ome/omero-web/omeroweb/settings.py
```

So I think that file needs `black` formatting applied.

**Comment by will-moore** [comment]
Testing locally, a quick browse around the webclient looks good. The major blocker on having this included in our daily merge-ci build is the `omero.web.apps` mentioned above.
The current list there is:
 - [ ] omero_figure
 - [ ] omero_iviewer
 - [ ] omero_parade
 - [ ] omero_fpbioimage
 - [ ] omero_gallery
 - [ ] omero_mapr
 - [ ] omero_weberror
 - [ ] omero_webtest
 - [ ] omero_webtagging_autotag
 - [ ] omero_webtagging_tagsearch
 - [ ] omero_openwith_url
 - [ ] parade_crossfilter (github.com/will-moore/parade-crossfilter)

We don't need to have all of these updated before we start including this PR in the merge build (we can temporarily comment-out the install of some) but at least figure & iviewer. Then follow up with mapr and gallery ASAP, then fpbioimage and the rest...

As with the last Django upgrade, this shouldn't take very long to tackle each app.

**Comment by bramalingam** [comment]
Thank you for the detailed review @will-moore. I will fix the black formatting and push a commit. Happy to create separate PR's for iViewer and figure as well once this is merged.  I still haven't finished the unit-tests, will wrap up whatever I have this weekend and push to this branch.

**Comment by bramalingam** [comment]
@will-moore : I've added a test that tests URLs which do not use any arguments, and covers a fraction of URLs. And I've fixed the black formatting as well. The test could be further improved to make sure URLs with arguments are handled as well. The test is just to make sure that I have not broken any URL patterns. I am happy to get started with the PR's for Figure & iviewer, if that sounds good to you.

**Comment by will-moore** [comment]
@bramalingam Sounds great. Figure and iviewer don't have so much Python code and not many URLs, so hopefully won't be much to fix. Thanks.

**Comment by bramalingam** [comment]
Hi @will-moore : I have opened up two PR's for figure and viewer, it would be great if they could be integrated with this PR and tested (locally) by someone. Am happy to open PR's for the other apps if all these changes look sensible and are not breaking any existing urls.

**Comment by chris-allan** [comment]
@will-moore: Propose that we close this PR in favour of #356. Can always reopen if required.

**Comment by jburel** [comment]
replaced by https://github.com/ome/omero-web/pull/356
